The job **`aarch64-gnu-llvm-21-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25278916601/job/74113052534) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/74113052534) [(plain)](https://github.com/rust-lang/rust/commit/f50085d01a00f06d4292fed607de1508b03c3af0/checks/74113052534/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Executing "/scripts/stage_2_test_set1.sh"
+ /scripts/stage_2_test_set1.sh
PR_CI_JOB set; skipping tidy
+ '[' 1 == 1 ']'
+ echo 'PR_CI_JOB set; skipping tidy'
+ SKIP_TIDY='--skip tidy'
+ ../x.py --stage 2 test --skip tidy --skip compiler --skip src
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2026-04-14/rustfmt-nightly-aarch64-unknown-linux-gnu.tar.xz
---
1 error[E0308]: mismatched types
-   --> $DIR/ref-pat-suggestions.rs:7:18
+   --> $DIR/ref-pat-suggestions.rs:6:18
3    |
4 LL | fn pin_mut_param(&pin mut x: i32) {}
5    |                  ^^^^^^^^^^  --- expected due to this

9    = note: expected type `i32`
10             found struct `Pin<&mut _>`
11 note: to declare a mutable parameter use: `mut x`
-   --> $DIR/ref-pat-suggestions.rs:7:18
+   --> $DIR/ref-pat-suggestions.rs:6:18
13    |
14 LL | fn pin_mut_param(&pin mut x: i32) {}
15    |                  ^^^^^^^^^^

20    |
21 
22 error[E0308]: mismatched types
-   --> $DIR/ref-pat-suggestions.rs:10:20
+   --> $DIR/ref-pat-suggestions.rs:9:20
24    |
25 LL | fn pin_const_param(&pin const x: i32) {}
26    |                    ^^^^^^^^^^^^  --- expected due to this

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/pin-ergonomics/ref-pat-suggestions.rs:6:18
-   --> /checkout/tests/ui/pin-ergonomics/ref-pat-suggestions.rs:6:18
---
3 #![feature(pin_ergonomics)]
- #![allow(incomplete_features)]
5 #![allow(unused)]
6 
7 fn pin_mut_param(x: &pin mut i32) {}


The actual fixed differed from the expected fixed
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args pin-ergonomics/ref-pat-suggestions.rs`

error: 2 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/pin-ergonomics/ref-pat-suggestions.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/pin-ergonomics/ref-pat-suggestions" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/pin-ergonomics/ref-pat-suggestions.rs:6:18
   |
LL | fn pin_mut_param(&pin mut x: i32) {}
   |                  ^^^^^^^^^^  --- expected due to this
   |                  |
   |                  expected `i32`, found `Pin<&mut _>`
   |
   = note: expected type `i32`
            found struct `Pin<&mut _>`
note: to declare a mutable parameter use: `mut x`
  --> /checkout/tests/ui/pin-ergonomics/ref-pat-suggestions.rs:6:18
   |
LL | fn pin_mut_param(&pin mut x: i32) {}
   |                  ^^^^^^^^^^
help: to take parameter `x` by reference, move `&pin mut` to the type
   |
LL - fn pin_mut_param(&pin mut x: i32) {}
LL + fn pin_mut_param(x: &pin mut i32) {}
   |

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/pin-ergonomics/ref-pat-suggestions.rs:9:20
   |
LL | fn pin_const_param(&pin const x: i32) {}
   |                    ^^^^^^^^^^^^  --- expected due to this
   |                    |
   |                    expected `i32`, found `Pin<&_>`
   |
   = note: expected type `i32`
            found struct `Pin<&_>`
help: to take parameter `x` by reference, move `&pin const` to the type
   |
LL - fn pin_const_param(&pin const x: i32) {}
LL + fn pin_const_param(x: &pin const i32) {}
   |

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
```

</details>

--- a/compiler/rustc_hir_typeck/src/pat.rs
+++ b/compiler/rustc_hir_typeck/src/pat.rs
@@ -1352,7 +1352,6 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     }
 
     /// Precondition: pat is a `Ref(_)` pattern
-    // FIXME(pin_ergonomics): add suggestions for `&pin mut` or `&pin const` patterns
     fn borrow_pat_suggestion(&self, err: &mut Diag<'_>, pat: &Pat<'_>) {
         let tcx = self.tcx;
         if let PatKind::Ref(inner, pinned, mutbl) = pat.kind
@@ -1407,6 +1406,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             };
 
             match binding_parent {
+                hir::Node::Param(hir::Param { ty_span, pat, .. })
+                    if pat.span != *ty_span
+                        && pinned.is_pinned()
+                        && !tcx.features().pin_ergonomics() =>
+                {
+                    // FIXME(pin_ergonomics): Once `pin_ergonomics` is stabilized, remove this
+                    // gate and allow the pinned reference type-position suggestion unconditionally.
+                }
                 // Check that there is explicit type (ie this is not a closure param with inferred type)
                 // so we don't suggest moving something to the type that does not exist
                 hir::Node::Param(hir::Param { ty_span, pat, .. }) if pat.span != *ty_span => {
@@ -1414,7 +1421,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         format!("to take parameter `{binding}` by reference, move `&{pin_and_mut}` to the type"),
                         vec![
                             (pat.span.until(inner.span), "".to_owned()),
-                            (ty_span.shrink_to_lo(), mutbl.ref_prefix_str().to_owned()),
+                            (ty_span.shrink_to_lo(), format!("&{}", pinned.prefix_str(mutbl))),
                         ],
                         Applicability::MachineApplicable
                     );

--- a/tests/ui/pin-ergonomics/ref-pat-suggestions.fixed
+++ b/tests/ui/pin-ergonomics/ref-pat-suggestions.fixed
@@ -1,0 +1,12 @@
+//@ run-rustfix
+
+#![feature(pin_ergonomics)]
+#![allow(unused)]
+
+fn pin_mut_param(x: &pin mut i32) {}
+//~^ ERROR mismatched types
+
+fn pin_const_param(x: &pin const i32) {}
+//~^ ERROR mismatched types
+
+fn main() {}

--- a/tests/ui/pin-ergonomics/ref-pat-suggestions.rs
+++ b/tests/ui/pin-ergonomics/ref-pat-suggestions.rs
@@ -1,0 +1,12 @@
+//@ run-rustfix
+
+#![feature(pin_ergonomics)]
+#![allow(unused)]
+
+fn pin_mut_param(&pin mut x: i32) {}
+//~^ ERROR mismatched types
+
+fn pin_const_param(&pin const x: i32) {}
+//~^ ERROR mismatched types
+
+fn main() {}

--- a/tests/ui/pin-ergonomics/ref-pat-suggestions.stderr
+++ b/tests/ui/pin-ergonomics/ref-pat-suggestions.stderr
@@ -1,0 +1,40 @@
+error[E0308]: mismatched types
+  --> $DIR/ref-pat-suggestions.rs:6:18
+   |
+LL | fn pin_mut_param(&pin mut x: i32) {}
+   |                  ^^^^^^^^^^  --- expected due to this
+   |                  |
+   |                  expected `i32`, found `Pin<&mut _>`
+   |
+   = note: expected type `i32`
+            found struct `Pin<&mut _>`
+note: to declare a mutable parameter use: `mut x`
+  --> $DIR/ref-pat-suggestions.rs:6:18
+   |
+LL | fn pin_mut_param(&pin mut x: i32) {}
+   |                  ^^^^^^^^^^
+help: to take parameter `x` by reference, move `&pin mut` to the type
+   |
+LL - fn pin_mut_param(&pin mut x: i32) {}
+LL + fn pin_mut_param(x: &pin mut i32) {}
+   |
+
+error[E0308]: mismatched types
+  --> $DIR/ref-pat-suggestions.rs:9:20
+   |
+LL | fn pin_const_param(&pin const x: i32) {}
+   |                    ^^^^^^^^^^^^  --- expected due to this
+   |                    |
+   |                    expected `i32`, found `Pin<&_>`
+   |
+   = note: expected type `i32`
+            found struct `Pin<&_>`
+help: to take parameter `x` by reference, move `&pin const` to the type
+   |
+LL - fn pin_const_param(&pin const x: i32) {}
+LL + fn pin_const_param(x: &pin const i32) {}
+   |
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

